session/session.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2018-2022 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package session
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"os/exec"
+
+	// We use this ssh because it implements port redirection.
+	// It can not, however, unpack password-protected keys yet.
+
+	"github.com/hashicorp/go-multierror" // TODO: get rid of krpty
+	"github.com/u-root/u-root/pkg/termios"
+	"golang.org/x/sys/unix"
+)
+
+// Bind defines a bind mount. It records the Local directory,
+// e.g. /bin, and the remote directory, e.g. /tmp/cpu/bin.
+type Bind struct {
+	Local  string
+	Remote string
+}
+
+// N.B. we used to have a Server type. But at some point I realized
+// that a cpu server is just an SSH server with a special handler.
+// Because one specializes the SSH server with cpu-specific attributes
+// such as port and handler, further containing that SSH server in a
+// CPU server led to awkwardness: the cpu Server struct contained
+// an SSH server struct which contained cpu-specific entities.
+// In the end, a CPU server is just an SSH server with some defaults
+// changed, so for now, we just make the cpu Server type an SSH
+// server type. Making our own type gives us an escape route
+// in the future should we realize we made a mistake.
+
+// Session is one instance of a cpu session, started by a cpud.
+type Session struct {
+	restorer *termios.Termios
+	Stdin    io.Reader
+	Stdout   io.Writer
+	Stderr   io.Writer
+	binds    []Bind
+	// Any function can use fail to mark that something
+	// went badly wrong in some step. At that point, if wtf is set,
+	// cpud will start it. This is incredibly handy for debugging.
+	fail   bool
+	msize  int
+	mopts  string
+	port9p string
+	cmd    string
+	args   []string
+}
+
+var (
+	v = func(string, ...interface{}){}
+	// To get debugging when Things Go Wrong, you can run as, e.g., -wtf /bbin/elvish
+	// or change the value here to /bbin/elvish.
+	// This way, when Things Go Wrong, you'll be dropped into a shell and look around.
+	// This is sometimes your only way to debug if there is (e.g.) a Go runtime
+	// bug around unsharing. Which has happened.
+	// This is compile time only because I'm so uncertain of whether it's dangerous
+	wtf string
+)
+
+// DropPrivs drops privileges to the level of os.Getuid / os.Getgid
+func (s *Session) DropPrivs() error {
+	uid := unix.Getuid()
+	v("CPUD:dropPrives: uid is %v", uid)
+	if uid == 0 {
+		v("CPUD:dropPrivs: not dropping privs")
+		return nil
+	}
+	gid := unix.Getgid()
+	v("CPUD:dropPrivs: gid is %v", gid)
+	if err := unix.Setreuid(-1, uid); err != nil {
+		return err
+	}
+	return unix.Setregid(-1, gid)
+}
+
+// Terminal sets up an interactive terminal.
+func (s *Session) Terminal() error {
+	// for some reason echo is not set.
+	t, err := termios.New()
+	if err != nil {
+		return fmt.Errorf("CPUD:can't get a termios; oh well; %v", err)
```

Comment:
To distinguish from the following message, this could be "CPUD: can't create...".

Reply:
on yeah this is a screwup, will fix.